### DISCUSSION TITLE
Can't Request Review From Team On GitHub From Forking Renovate

### DISCUSSION THREAD

**Discussion by Kurt-von-Laven** [renovatebot/renovate#24226] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Renovate requests review of `@ScribeMD/dependency-management` on all pull requests it opens.

### What you tried so far.

We [extend `:reviewer(team:dependency-management)`](https://github.com/ScribeMD/.github/blob/6cb1ed28eed5c20c0c5c8ffdb008b9424ef58f68/default.json#L7) in our base Renovate config based on the following excerpt from [the official docs](https://docs.renovatebot.com/configuration-options/#reviewers):

> If you're assigning a team to review on GitHub, you must use the prefix `team:` and add the last part of the team name.

We successfully manually requested review of `@ScribeMD/dependency-management` on a pull request Renovate opened. Could this be failing on account of some permission that Forking Renovate doesn't have but Mend Renovate does?

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Adding reviewers 'team:dependency-management' to #191 (branch="renovate/frnmst-md-toc-8.x")
DEBUG: POST https://api.github.com/repos/ScribeMD/template-repo/pulls/191/requested_reviewers = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=422 retryCount=0, duration=199) (branch="renovate/frnmst-md-toc-8.x")
DEBUG: 422 Error thrown from GitHub (branch="renovate/frnmst-md-toc-8.x")
{
  "err": {
    "name": "HTTPError",
    "code": "ERR_NON_2XX_3XX_RESPONSE",
    "timings": {
      "start": 1693798732147,
      "socket": 1693798732147,
      "lookup": 1693798732148,
      "connect": 1693798732150,
      "secureConnect": 1693798732154,
      "upload": 1693798732154,
      "response": 1693798732346,
      "end": 1693798732346,
      "phases": {
        "wait": 0,
        "dns": 1,
        "tcp": 2,
        "tls": 4,
        "request": 0,
        "firstByte": 192,
        "download": 0,
        "total": 199
      }
    },
    "message": "Response code 422 (Unprocessable Entity)",
    "stack": "HTTPError: Response code 422 (Unprocessable Entity)\n    at Request.<anonymous> (/opt/containerbase/tools/renovate/36.78.8/node_modules/got/dist/source/as-promise/index.js:118:42)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)",
    "options": {
      "headers": {
        "user-agent": "RenovateBot/36.78.8 (https://github.com/renovatebot/renovate)",
        "accept": "application/json, application/vnd.github.machine-man-preview+json",
        "authorization": "***********",
        "content-type": "application/json",
        "content-length": "59",
        "accept-encoding": "gzip, deflate, br"
      },
      "url": "https://api.github.com/repos/ScribeMD/template-repo/pulls/191/requested_reviewers",
      "hostType": "github",
      "username": "",
      "password": "",
      "method": "POST",
      "http2": false
    },
    "response": {
      "statusCode": 422,
      "statusMessage": "Unprocessable Entity",
      "body": {
        "message": "Validation Failed",
        "errors": [
          "Could not resolve to a node with the global id of 'T_kwDOBKjlfs4AggHY'."
        ],
        "documentation_url": "https://docs.github.com/rest/pulls/review-requests#request-reviewers-for-a-pull-request"
      },
      "headers": {
        "server": "GitHub.com",
        "date": "Mon, 04 Sep 2023 03:38:52 GMT",
        "content-type": "application/json; charset=utf-8",
        "content-length": "226",
        "x-github-media-type": "github.v3; param=machine-man-preview",
        "x-accepted-github-permissions": "pull_requests=write",
        "x-github-api-version-selected": "2022-11-28",
        "x-ratelimit-limit": "5000",
        "x-ratelimit-remaining": "4981",
        "x-ratelimit-reset": "1693801152",
        "x-ratelimit-used": "19",
        "x-ratelimit-resource": "core",
        "access-control-expose-headers": "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset",
        "access-control-allow-origin": "*",
        "strict-transport-security": "max-age=31536000; includeSubdomains; preload",
        "x-frame-options": "deny",
        "x-content-type-options": "nosniff",
        "x-xss-protection": "0",
        "referrer-policy": "origin-when-cross-origin, strict-origin-when-cross-origin",
        "content-security-policy": "default-src 'none'",
        "vary": "Accept-Encoding, Accept, X-Requested-With",
        "x-github-request-id": "BED1:87FB:187D4C1:327BDF8:64F5514C",
        "connection": "close"
      },
      "httpVersion": "1.1",
      "retryCount": 0
    }
  }
}

WARN: Failed to assign reviewer (branch="renovate/frnmst-md-toc-8.x")
{
  "err": {
    "hostType": "github",
    "err": {
      "name": "HTTPError",
      "code": "ERR_NON_2XX_3XX_RESPONSE",
      "timings": {
        "start": 1693798732147,
        "socket": 1693798732147,
        "lookup": 1693798732148,
        "connect": 1693798732150,
        "secureConnect": 1693798732154,
        "upload": 1693798732154,
        "response": 1693798732346,
        "end": 1693798732346,
        "phases": {
          "wait": 0,
          "dns": 1,
          "tcp": 2,
          "tls": 4,
          "request": 0,
          "firstByte": 192,
          "download": 0,
          "total": 199
        }
      },
      "message": "Response code 422 (Unprocessable Entity)",
      "stack": "HTTPError: Response code 422 (Unprocessable Entity)\n    at Request.<anonymous> (/opt/containerbase/tools/renovate/36.78.8/node_modules/got/dist/source/as-promise/index.js:118:42)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)",
      "options": {
        "headers": {
          "user-agent": "RenovateBot/36.78.8 (https://github.com/renovatebot/renovate)",
          "accept": "application/json, application/vnd.github.machine-man-preview+json",
          "authorization": "***********",
          "content-type": "application/json",
          "content-length": "59",
          "accept-encoding": "gzip, deflate, br"
        },
        "url": "https://api.github.com/repos/ScribeMD/template-repo/pulls/191/requested_reviewers",
        "hostType": "github",
        "username": "",
        "password": "",
        "method": "POST",
        "http2": false
      },
      "response": {
        "statusCode": 422,
        "statusMessage": "Unprocessable Entity",
        "body": {
          "message": "Validation Failed",
          "errors": [
            "Could not resolve to a node with the global id of 'T_kwDOBKjlfs4AggHY'."
          ],
          "documentation_url": "https://docs.github.com/rest/pulls/review-requests#request-reviewers-for-a-pull-request"
        },
        "headers": {
          "server": "GitHub.com",
          "date": "Mon, 04 Sep 2023 03:38:52 GMT",
          "content-type": "application/json; charset=utf-8",
          "content-length": "226",
          "x-github-media-type": "github.v3; param=machine-man-preview",
          "x-accepted-github-permissions": "pull_requests=write",
          "x-github-api-version-selected": "2022-11-28",
          "x-ratelimit-limit": "5000",
          "x-ratelimit-remaining": "4981",
          "x-ratelimit-reset": "1693801152",
          "x-ratelimit-used": "19",
          "x-ratelimit-resource": "core",
          "access-control-expose-headers": "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset",
          "access-control-allow-origin": "*",
          "strict-transport-security": "max-age=31536000; includeSubdomains; preload",
          "x-frame-options": "deny",
          "x-content-type-options": "nosniff",
          "x-xss-protection": "0",
          "referrer-policy": "origin-when-cross-origin, strict-origin-when-cross-origin",
          "content-security-policy": "default-src 'none'",
          "vary": "Accept-Encoding, Accept, X-Requested-With",
          "x-github-request-id": "BED1:87FB:187D4C1:327BDF8:64F5514C",
          "connection": "close"
        },
        "httpVersion": "1.1",
        "retryCount": 0
      }
    },
    "message": "external-host-error",
    "stack": "Error: external-host-error\n    at handleGotError (/opt/containerbase/tools/renovate/36.78.8/node_modules/renovate/lib/util/http/github.ts:149:12)\n    at GithubHttp.request (/opt/containerbase/tools/renovate/36.78.8/node_modules/renovate/lib/util/http/github.ts:371:13)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at GithubHttp.requestJson (/opt/containerbase/tools/renovate/36.78.8/node_modules/renovate/lib/util/http/index.ts:289:17)\n    at Proxy.addReviewers (/opt/containerbase/tools/renovate/36.78.8/node_modules/renovate/lib/modules/platform/github/index.ts:1308:5)\n    at addParticipants (/opt/containerbase/tools/renovate/36.78.8/node_modules/renovate/lib/workers/repository/update/pr/participants.ts:88:11)\n    at ensurePr (/opt/containerbase/tools/renovate/36.78.8/node_modules/renovate/lib/workers/repository/update/pr/index.ts:502:9)\n    at processBranch (/opt/containerbase/tools/renovate/36.78.8/node_modules/renovate/lib/workers/repository/update/branch/index.ts:750:28)\n    at writeUpdates (/opt/containerbase/tools/renovate/36.78.8/node_modules/renovate/lib/workers/repository/process/write.ts:159:17)\n    at update (/opt/containerbase/tools/renovate/36.78.8/node_modules/renovate/lib/workers/repository/process/extract-update.ts:214:11)\n    at Object.renovateRepository (/opt/containerbase/tools/renovate/36.78.8/node_modules/renovate/lib/workers/repository/index.ts:76:19)\n    at attributes.repository (/opt/containerbase/tools/renovate/36.78.8/node_modules/renovate/lib/workers/global/index.ts:184:11)\n    at start (/opt/containerbase/tools/renovate/36.78.8/node_modules/renovate/lib/workers/global/index.ts:169:7)\n    at /opt/containerbase/tools/renovate/36.78.8/node_modules/renovate/lib/renovate.ts:18:22"
  }
}

DEBUG: Added reviewers (branch="renovate/frnmst-md-toc-8.x")
{
  "reviewers": [
    "team:dependency-management"
  ]
}
```

</details>

**Comment by rarkins** [comment]
There doesn't seem to be any difference in permissions between the apps which should make any difference. Do you _know_ that the main app works for this or are you just assuming it does? The error from GitHub is pretty unhelpful: `"Could not resolve to a node with the global id of 'T_kwDOBKjlfs4AggHY'."`

**Reply by Kurt-von-Laven** [reply]
I was just guessing there would have been a lot more reports about this if it didn't work for the main app. Yeah, I saw that lovely gem in the logs too. I think it may be trying to say that it couldn't find that ID, but without more logging I would have to guess what it's the ID of.

**Reply by rarkins** [reply]
I've reviewed the permissions plus the API docs and I can't see anything obvious which could be wrong. I agree with you that we'd likely have heard about it if teams assigning didn't work for anyone.

**Reply by Kurt-von-Laven** [reply]
[rowi1de/auto-assign-review-teams#17](https://github.com/rowi1de/auto-assign-review-teams/issues/17#issuecomment-1202884997) describes a non-obvious permission issue that could be the issue. If that is the problem, I suppose the docs could simply refer users of Forking Renovate to `CODEOWNERS`?

**Reply by rarkins** [reply]
That post mentions "...permissions to add teams as reviewers". I searched further and found this post: https://github.com/orgs/community/discussions/50704#discussioncomment-5393104

This one claims the token needs the "admin:org" permission. That appears to be documented here: https://docs.github.com/en/rest/overview/permissions-required-for-github-apps?apiVersion=2022-11-28#organization-administration

Not sure if it requires read or write access. Very few would want to give write access, so we wouldn't ask for that. Many might be concerned about read access too

I'd guess it's the https://docs.github.com/en/rest/teams/teams?apiVersion=2022-11-28#list-teams endpoint which is required, which falls under "members": https://docs.github.com/en/rest/overview/permissions-required-for-github-apps?apiVersion=2022-11-28#organization-permissions-for-members

**Reply by rarkins** [reply]
But Forking Renovate already has that permission: 
<img width="747" alt="image" src="https://github.com/renovatebot/renovate/assets/6311784/3b7210a3-f14c-4e04-95a7-81f84f05cd60">

**Reply by Kurt-von-Laven** [reply]
Thank you for all of the helpful links and jumping on this so quickly! I suspect the issue is that Mend Renovate presumably has `pull-requests: write`, but Forking Renovate doesn't, and [this endpoint](https://docs.github.com/en/rest/pulls/review-requests?apiVersion=2022-11-28#request-reviewers-for-a-pull-request), which is definitely required, [requires it](https://docs.github.com/en/rest/overview/permissions-required-for-github-apps?apiVersion=2022-11-28#repository-permissions-for-pull-requests).

**Reply by rarkins** [reply]
Forking Renovate already has this: 
<img width="739" alt="image" src="https://github.com/renovatebot/renovate/assets/6311784/af813617-2e45-4ccb-a9c7-8c676d0ef38f">

**Reply by Kurt-von-Laven** [reply]
Shoot. I took a closer look at https://github.com/orgs/community/discussions/50704#discussioncomment-5393104, and I believe that thread is unrelated, because they are talking about adding a team as a maintainer of a repository rather than adding a team as a reviewer of a pull request. I can try using a simple GitHub Actions workflow to make a request to the same endpoint I just linked to sort out which permissions are required (or possibly prove that even with all the permissions, it still doesn't work).

**Reply by Kurt-von-Laven** [reply]
I got [the same error message (including the same ID)](https://github.com/ScribeMD/slack-templates/actions/runs/6070948686/job/16468001352) trying to request review from the Dependency Management team in a different repository even with [`permissions: write-all`](https://github.com/ScribeMD/slack-templates/actions/runs/6070948686/workflow#L6). I assume, therefore, that the ID is that of the Dependency Management team. There are some permissions that are off limits to GitHub Actions for security reasons though, and I haven't yet tested with a PAT.

[GitHub's docs](https://docs.github.com/en/rest/pulls/review-requests?apiVersion=2022-11-28#request-reviewers-for-a-pull-request--status-codes) say a 422 means the user is not a collaborator, which is contradicted by the fact that we are able to request review from the team manually.

**Reply by Kurt-von-Laven** [reply]
I had no problems requesting review of the Dependency Management team using a PAT (Personal Access Token) with only `pull-requests: write` and the mandatory `metadata: read` permission. I guess it's conceivable that the Forking Renovate app could need `metadata: read`?

**Reply by rarkins** [reply]
`Metadata` (which I assume is the same thing for apps) is a mandatory access which all apps have:

![image](https://github.com/renovatebot/renovate/assets/6311784/f6e4c15f-124c-4d10-aa5f-72ff1aa4b9b8)

**Reply by Kurt-von-Laven** [reply]
In that case, it sounds like this is a bug I should report to GitHub?

**Reply by Kurt-von-Laven** [reply]
Thanks for all the help. I filed [orgs/community#66049](https://github.com/orgs/community/discussions/66049) linking to this discussion. I'll follow up here when the upstream issue is fixed to confirm that this feature is working in Renovate. I'm increasingly suspicious that this feature doesn't work in the Mend Renovate GitHub App either since I encountered the same issue curling the same REST endpoint from GitHub Actions. It's possible even though I have never seen this feature working that it was only broken recently by an upstream change.
